### ISSUE TITLE
Histogram latencies does not match with real latencies?

### ISSUE THREAD

**Issue by ligreman** [Kong/kong-plugin-prometheus#136]
### Use case

We have several services that expose APIs. We are testing Kong with one of these services.

I've created a Route + Service + Upstream (tested with round robin and least-connections) + 2 targets (2 instances of the API server).

We want to monitor Kong, using Prometheus + Grafana. Specially the requests latencies.

### Environment used

We have these machines, all with Ubuntu 20.04:
- 2 Kong nodes
  - Last version of kong v2.4.1
  - Prometheus plugin enabled globally (the plugin version that kong integrates)
  - There is a Nginx that balances requests between the 2 nodes, with round-robin 50-50
- Machine with Prometheus v2.15.2
  - Scrapes kong metrics every 15 seconds
- Machine with Grafana (kong official dashboard updated this week)

### Test executed

We launch a client (automatic process) that generates requests, aprox 10 request per second. 100k requests along several hours.

We monitor the requests via Kong, in the client that generates the requests and inside the API servers too.

### Behaviour detected

In Grafana we get **upstream** latencies around 30 seconds (between 25 and 35). Kong latencies are 10-15 milliseconds.

The times measured in the **API servers** (we measure the time between receiving the request and sending the response) are about half of that times, 10~15 seconds. This is the real time spent in the API generating and sending the response.

The times measured in the **client** that generates the requests are similar, 11~16 seconds. We measure the time between sending the request and receiving the response. So this is the real total time elapsed.

In the responses obtained, we have checked the latencies headers ('x-kong-proxy-latency' and 'x-kong-upstream-latency'). The upstream latency here in every response is very similar to the ones measured in the API servers, 10~15 seconds. They don't match the ones in the prometheus plugin.

### Possible bug??

**We don't understand that difference in latencies**. We know that histogram latencies are estimates (not exact times) and have some margin of error, but that much (the double)?

Or are we interpreting the latencies in the wrong way? (in that case, what indicates those latencies?)

Thanks for your attention.

**Comment by hbagdi** [comment]
I'm assuming that you are looking at quantile values in Grafana that are obtained by munging data in histogram timeseries stored in Prometheus. Is that accurate?

Assuming yes, the reason for this:
- Quantile estimations using histograms are not accurate.
- The are more accurate if the histogram buckets are configured properly
- In our case, the bucket values are [hard-coded](https://github.com/Kong/kong-plugin-prometheus/blob/master/kong/plugins/prometheus/exporter.lua#L15-L17) and they are hard-coded with an assumption that upstream services response times are in the order of milliseconds or a only a few seconds. In your case, that is probably what is causing the issue because we don't have nearly as many buckets in 5000 to 60000 ms range.

We could make the buckets configurable but that's putting more responsibility on the end user and in this case, most users don't have enough background to configure the buckets correctly.

**Comment by ligreman** [comment]
@hbagdi Yes, your assumtion is correct (sorry if my explanation is not very accurate, I don't have many experience with Grafana and Prometheus, and the buckets).

I understand that adding bucket configuration to the plugin could be not very useful for most users. I leave here an idea: 

> Having 2 or 3 sets of predefined buckets to select from, with a "human readable" label, like "(Default) Expected latency ranges from 1ms to 500ms", "Expected latency ranges from 1ms to 10000ms", "Expected latency ranges from 1ms to 30000ms".

Anyway, as a test I could try to fork the plugin and modify the values and then deploy the plugin in our Kong installation as a prometheus-custom one, I've done that with other Kong plugin before.

I am one of those users which don't have enough background metrics and buckets (still learning...), so I really appreciate your help. The estimations could it be improved by adding more values to the default buckets?

`local DEFAULT_BUCKETS = { 1, 2, 5, 7, 10, 15, 20, 25, 30, 40, 50, 60, 70, 80, 90, 100, 200, 300, 400, 500, 1000, 2000, 5000, 10000, 30000, 60000 }`

What could be the solution? Adding more buckets like.. 6000, 7000, 8000...? Is there anything else to modify?

I suppose that the last version of kong-plugin-prometheus is compatible with the last version of Kong (2.4.1 right now), is that so?

Thank you so much for your help!

**Comment by ligreman** [comment]
I have forked the plugin adding more buckets and it improves the estimates, without affecting the performance or resource consumption notably.

So I'll use this modified version for a while to better test it, and if it performs correctly we will use it.

I modified them like this, to better match our needs:

`local DEFAULT_BUCKETS = { 1, 2, 3, 5, 7, 10, 15, 20, 25, 30, 40, 50, 60, 70, 80, 90, 100, 200, 300, 400, 500, 750, 1000, 1500, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 12500, 15000, 17500, 20000, 25000, 30000, 35000, 40000, 50000, 60000 }`

Sidenote: I tried to make the buckets "user configurable" but could not think a way to make it a configuration field that can be passed to the init_worker in "exporter.lua" (in the init_worker phase you cannot read the plugin configuration...), so I just modified the default values for now.

**Comment by hbagdi** [comment]
You are on the right track.
Increasing the number of buckets will increase the memory consumption of Kong and increase the number of time-series in Prometheus.

We would like to make these configurable but haven't figured out a way to do this yet. @fffonion might have some answers here.

**Comment by fffonion** [comment]
Yeah having a configurable bucket value is tricky as the way `prometheus` initializes. It's also kind of an global instance
per Kong cluster and let user to configure individually will mean having multiple `prometheus` instances, which won't
work well.
I think to make this to work, we will need to seperate the initialization of the prometheus library. For example, those timer
init are global, then you can have multiple prometheus instances, they don't manage timers but just stuff like buckets and namespaces.
